### DISCUSSION TITLE
Error dialog on startup and app window position

### DISCUSSION THREAD

**Discussion by dlek** [beeware/toga#1433] [Q&A] [answered]
Hello community.  I have a small utility using Toga and it's working well but I am having two problems polishing it off.

1. If there is an error on startup, I would like to display an error dialog and then quit.  However if I set up a dialog as a child of the main application window during the startup handler, the dialog and app stay in the background.  The best solution I have found is to set up a separate application window just for the error dialog, but it's inconsistent in presentation with the standard `error_dialog()` result, and really seems like the wrong approach.
2. I have been unable to figure out how to position the (main) window so it is centered.  (Actually, setting the position appears to have no effect at this point, but it's moot unless I can determine where to place the window.)  I cannot find a way to query the geometry of the desktop, nor can I find a window placement hint mechanism (something like `placement=toga.layout.CENTRE`).  Dialogs are centered, but looking at the Toga source for Cocoa, not that I know anything about Cocoa dev, it looks like Toga makes a call to the OS libraries to create dialogs.  Is there a way to centre the application window?

Thanks in advance for any help or guidance!

**Comment by samschott** [comment, marked as answer]
I can definitely answer the second question and I'm afraid this is not currently possible. As you have discovered, the `Window.position` property currently does not do anything. PR #1395 will at least change this for macOS when it lands. There is also currently no toga API to query the desktop geometry or to directly center a window, you'll need to fall back to native APIs of each platform at the moment.

Regarding the first question, the behaviour here may depend on the platform implementation and at which point during the startup the app gains focus and is ordered to the front. Which platform are you trying this on?

**Reply by dlek** [reply]
Thanks for your reply @samschott.

Regarding centring the window, I'll watch that PR and see what I can figure out about the native API for Cocoa.

For the first question, this is also Cocoa/Mac OS.  The utility basically presents a single dialog allowing the user to set a couple of values which then sends an API request to an online service.  If the request fails, the user gets an error dialog using `app.main_window.error_dialog()`.  If the request succeeds, the app exits gracefully.  Very simple.  The request failure dialog works exactly as expected and desired.  The problem is if the app's configuration is missing or wrong or the invocation is bad: at that stage the application window has not initialized; this is detected before the event loop is initialized.  To borrow from the first example in the documentation:

```
# imports and the rest of the example
# ...

runtime = {}

def build(app):
    if runtime['fatal']:
        app.main_window.error_dialog("Fatal", f"Ack!  Pth!  Because {runtime['fatal']}")
        app.exit()
    # do the good stuff
    # ...

def init():
    try:
        runtime['config'] = configmodule.get_my_config()
    except ConfigError as e:
        runtime['fatal'] = e

def main():
    init()

    return toga.App(
        'First App',
        'org.beeware.helloworld',
        author='Tiberius Yak',
        description="A testing app",
        version=__version__,
        home_page="https://beeware.org",
        startup=build
    )

if __name__ == '__main__':
    main().main_loop()
```

So what happens here is the app never comes to the foreground in a failed-initialization case.  Launched from a terminal window, I get a bouncing icon in the Dock and have to switch to it to see the error message.  (It only just occurred to me now that launching from the terminal window might be part of the problem but the app works as expected otherwise.)

The problem seems to be that the event loop isn't in play yet because the app startup hasn't returned yet.  So I've tried to think of ways to delay launching the dialog until after the startup function has returned, but so far, the best I have come up with is using a different startup to build my own error dialog, and it's inconsistent with the other.

Not ideal but acceptable for what this is at this stage, but would be nice to know if there's a better way.

**Reply by dlek** [reply]
I've [posted a gist](https://gist.github.com/dlek/abbfc19fbca735f498a7c7cde075f1c0) demonstrating the issue and showing a few of the things I've tried to solve it/get around it.

**Reply by samschott** [reply]
Thanks for the additional details. Yes, the problem is indeed that you show the error dialog before the event loop is initialized. One way around that would be to schedule the execution of a method which show the error dialog *after* startup has completed. You can use `toga.TogaApp.add_background_task()` to do this, for example:

```Python
def build(app):
    if runtime['fatal']:
        app.add_background_task(show_error_and_exit)
    else:
        # do the good stuff
        # ...

def show_error_and_exit(sender):
    sender.main_window.error_dialog("Fatal", f"Ack!  Pth!  Because {runtime['fatal']}")
    sender.exit()
```

This will unfortunately currently fail as well, because the platform implementation instead of the TogaApp is passed to the background task. I've submitted #1435 to fix this. In the meantime, use this workaround to get back to the interface:

```Python
def show_error_and_exit(sender):
    sender.interface.main_window.error_dialog("Fatal", f"Ack!  Pth!  Because {runtime['fatal']}")
    sender.interface.exit()
```

**Reply by dlek** [reply]
Thank you so much for your help!  I've implemented this and in works as I've hoped, and I've added a TODO in the code to update when #1435 is merged into a release.

Since I've asked two questions I can probably only accept one.  At any rate thanks again for your responsiveness.  It's a goofy utility that'll probably never be used by anyone other than me, but it's nice to get it right. :)
